internal/core/textlayout.rs

```diff
@@ -0,0 +1,954 @@
+
+                    if self.fragment_fits() {
+                        self.commit_fragment();
+                    } else {
+                        line_to_emit = Some(core::mem::take(&mut self.current_line));
+                        self.commit_fragment();
+                    }
+
+                    self.fragment.add_grapheme(&grapheme);
+                }
+                _ => {
+                    self.fragment.add_grapheme(&grapheme);
+
+                    if self.current_line_fits() && !self.fragment_fits() {
+                        if !self.current_line.byte_range.is_empty() {
+                            line_to_emit = Some(core::mem::take(&mut self.current_line));
+                        }
+                        self.commit_fragment();
+                    }
+                }
+            };
+
+            if line_to_emit.is_some() {
+                self.num_emitted_lines += 1;
+                return line_to_emit;
+            }
+        }
+
+        self.commit_fragment();
+        if !self.current_line.byte_range.is_empty() || self.num_emitted_lines == 0 {
+            self.num_emitted_lines += 1;
+            return Some(core::mem::take(&mut self.current_line));
+        }
+
+        None
+    }
+}
+
+// Measures the size of the given text when rendered with the specified font and optionally constrained
+// by the provided `max_width`.
+// Returns a tuple of the width of the longest line as well as the number of lines.
+pub fn text_size<Font: TextShaper>(
+    font: &Font,
+    text: &str,
+    max_width: Option<Font::Length>,
+) -> (Font::Length, Font::LengthPrimitive)
+where
+    Font::Length: core::fmt::Debug,
+{
+    let mut max_line_width = Font::Length::zero();
+    let mut line_count = Font::LengthPrimitive::zero();
+    let glyphs = RefCell::new(Vec::new());
+
+    for line in TextLineBreaker::new(text, font, &glyphs, max_width) {
+        max_line_width = euclid::approxord::max(max_line_width, line.text_width);
+        line_count += Font::LengthPrimitive::one();
+    }
+
+    (max_line_width, line_count)
+}
+
+/// Layout the given string in lines, and call the `layout_line` callback with the line to draw at position y.
+/// The signature of the `layout_line` function is: `(canvas, text, pos, start_index, line_metrics)`.
+/// start index is the starting byte of the text in the string.
+/// Returns the baseline y coordinate.
+pub fn layout_text_lines<Font: TextShaper>(
+    string: &str,
+    font: &Font,
+    font_height: Font::Length,
+    max_width: Font::Length,
+    max_height: Font::Length,
+    (horizontal_alignment, vertical_alignment): (TextHorizontalAlignment, TextVerticalAlignment),
+    wrap: TextWrap,
+    overflow: TextOverflow,
+    single_line: bool,
+    mut layout_line: impl FnMut(
+        &mut dyn Iterator<Item = (Font::Length, &'_ Font::Glyph)>,
+        Font::Length,
+        Font::Length,
+        &TextLine<Font::Length>,
+    ),
```

Comment:
I made it a line callback in the GL crate because the femtovg draw API draw lines.

We could also consider a callback per glyph instead of per line.
So that would be like:

```suggestion
    mut layout_glyph: impl FnMut(
        &'_ Font::Glyph,
        Font::Length, // x
        Font::Length, // y
        &TextLine<Font::Length>, // not sure about that one, is that not just &str?
    ),
```

Reply:
Yes, in principle it could also be per-glyph. That said, in the common case we already have the glyphs all ready for rendering in the glyph_buffer Vec, so it seems sensible to pass them on in batches as large as possible.